analyzer/emeraldtokens/emerald_tokens.go

```diff
@@ -0,0 +1,214 @@
+	ctx := context.Background()
+
+	// Initialize source storage.
+	networkCfg := oasisConfig.Network{
+		ChainContext: cfg.ChainContext,
+		RPC:          cfg.RPC,
+	}
+	factory, err := oasis.NewClientFactory(ctx, &networkCfg, cfg.FastStartup)
+	if err != nil {
+		logger.Error("error creating client factory",
+			"err", err,
+		)
+		return nil, err
+	}
+
+	network, err := analyzer.FromChainContext(cfg.ChainContext)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := analyzer.RuntimeEmerald.ID(network)
+	if err != nil {
+		return nil, err
+	}
+	logger.Info("Emerald runtime ID determined", "runtime_id", id)
+
+	client, err := factory.Runtime(id)
+	if err != nil {
+		logger.Error("error creating runtime client",
+			"err", err,
+		)
+		return nil, err
+	}
+
+	return &Main{
+		source: client,
+		target: target,
+		qf:     analyzer.NewQueryFactory("oasis_3", analyzer.RuntimeEmerald.String()),
+		logger: logger.With("analyzer", EmeraldDamaskTokensAnalyzerName),
+	}, nil
+}
+
+type StaleToken struct {
+	Addr                  string
+	LastMutateRound       uint64
+	Type                  *modules.EVMTokenType
+	AddrContextIdentifier string
+	AddrContextVersion    int
+	AddrData              []byte
+}
+
+func (m Main) getStaleTokens(ctx context.Context, limit int) ([]*StaleToken, error) {
+	var staleTokens []*StaleToken
+	rows, err := m.target.Query(ctx, m.qf.RuntimeEVMTokensAnalysisStaleQuery(), limit)
+	if err != nil {
+		return nil, fmt.Errorf("querying discovered tokens: %w", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var staleToken StaleToken
+		if err = rows.Scan(
+			&staleToken.Addr,
+			&staleToken.LastMutateRound,
+			&staleToken.Type,
+			&staleToken.AddrContextIdentifier,
+			&staleToken.AddrContextVersion,
+			&staleToken.AddrData,
+		); err != nil {
+			return nil, fmt.Errorf("scanning discovered token: %w", err)
+		}
+		staleTokens = append(staleTokens, &staleToken)
+	}
+	return staleTokens, nil
+}
+
+func (m Main) processBatch(ctx context.Context) error {
+	// This part is not in the update transaction. We don't want to hold
+	// up the tokens table while we download the token info. Make sure the
+	// Emerald analyzer and this Emerald tokens analyzer can cooperate
+	// correctly without stopping each other.
+	staleTokens, err := m.getStaleTokens(ctx, MaxDownloadBatch)
+	if err != nil {
+		return fmt.Errorf("getting discovered tokens: %w", err)
+	}
+	m.logger.Info("processing", "num_stale_tokens", len(staleTokens))
+
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, DownloadTimeout)
+	defer cancel()
+	group, groupCtx := errgroup.WithContext(ctxWithTimeout)
+
+	batch := &storage.QueryBatch{}
+
+	for _, st := range staleTokens {
+		// Redeclare `st` for unclobbered use within goroutine.
+		staleToken := st
+		group.Go(func() error {
+			m.logger.Info("downloading", "stale_token", staleToken)
+			// todo: assert that addr context is secp256k1
+			if staleToken.Type == nil {
```

Comment:
this condition isn't right. if the block scanner finds a possible token and we scan and find it's not a recognized token type, we won't create a row in evm_tokens. thus staleToken.Type will still be nil. should change this to query in the last_download_round and check if that's nil instead.

Reply:
fixed